### PR TITLE
Migrate pickled data & change XCom value type to JSON

### PR DESCRIPTION
**Pull request by kaxil**
follow-up of https://github.com/apache/airflow/pull/43905

Changes:
- Changed `XCom.value` column to JSON for all dbs.
- Archived pickled XCom data to `_xcom_archive` and removed it from the `xcom` table.
- Removed encoded string in XCom serialization and deserialization logic.
- Updated logic for `XComObjectStorageBackend` to make it compatible for AF 2 & 3

image

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/migrations/versions/0049_3_0_0_remove_pickled_data_from_xcom_table.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        "mysql": "HEX(value) LIKE '80%'",
+        "sqlite": "substr(value, 1, 1) = char(128)",
+    }
+
+    condition = condition_templates.get(dialect)
+    if not condition:
+        raise RuntimeError(f"Unsupported dialect: {dialect}")
+
+    # Key is a reserved keyword in MySQL, so we need to quote it
+    quoted_key = "`key`" if dialect == "mysql" else '"key"'
```
Comment (ashb) [inline airflow/migrations/versions/0049_3_0_0_remove_pickled_data_from_xcom_table.py]: This feels suspect -- like we should be able to use SQLA or alchemy to do this more safely? I guess it's not user input so we don't _have_ to.

File: airflow/migrations/versions/0049_3_0_0_remove_pickled_data_from_xcom_table.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        sa.Column("run_id", StringID(length=250), nullable=False),
+        sa.Column("value", sa.LargeBinary().with_variant(LONGBLOB, "mysql"), nullable=True),
+        sa.Column("timestamp", TIMESTAMP(), nullable=False),
+        sa.PrimaryKeyConstraint("dag_run_id", "task_id", "map_index", "key"),
+    )
+
+    # Condition to detect pickled data for different databases
+    condition_templates = {
+        "postgresql": "get_byte(value, 0) = 128",
+        "mysql": "HEX(value) LIKE '80%'",
```
Comment (ashb) [inline airflow/migrations/versions/0049_3_0_0_remove_pickled_data_from_xcom_table.py]: If this turns the entire blob into hex first, we should probably substring it first to not make it convert a "large" amount of data to hex

File: airflow/models/xcom.py [hunk 3]
```diff
@@ -80,7 +79,7 @@ class BaseXCom(TaskInstanceDependencies, LoggingMixin):
     dag_id = Column(String(ID_LEN, **COLLATION_ARGS), nullable=False)
     run_id = Column(String(ID_LEN, **COLLATION_ARGS), nullable=False)
 
-    value = Column(LargeBinary().with_variant(LONGBLOB, "mysql"))
+    value = Column(JSON)
```
Comment (ashb) [inline airflow/models/xcom.py:82]: Should this be non-NULL?

**Review by ashb — COMMENTED** [review]
I don't love the migration, but it seems "safe" enough (modulo the comment about `hex()` call on mysql path, and I can't think of a better approach either)

File: airflow/migrations/versions/0049_3_0_0_remove_pickled_data_from_xcom_table.py (hunk 2 above)
Comment (kaxil) [reply airflow/migrations/versions/0049_3_0_0_remove_pickled_data_from_xcom_table.py]: good point, fixed in https://github.com/apache/airflow/pull/44166/commits/e38956e5cc21334aa50babc569f9d2cb7d24d897

File: airflow/migrations/versions/0049_3_0_0_remove_pickled_data_from_xcom_table.py (hunk 1 above)
Comment (kaxil) [reply airflow/migrations/versions/0049_3_0_0_remove_pickled_data_from_xcom_table.py]: fixed in https://github.com/apache/airflow/pull/44166/commits/e38956e5cc21334aa50babc569f9d2cb7d24d897

File: airflow/models/xcom.py (hunk 3 above)
Comment (kaxil) [reply airflow/models/xcom.py:82]: It is handled in deserialization so its fine I think

https://github.com/apache/airflow/blob/e8fe1bddf1bf8ed4c4ce94fa6efd713be02fb6bc/airflow/models/xcom.py#L460-L469

**Review by ashb — APPROVED** [review]

**Comment by kaxil** [comment]
Static failure is unrelated

**Comment by jscheffl** [comment]
Thanks for the PR - I LIKE it! Sorry entering too late for review :-(

**Pull request merged**